### PR TITLE
mgr,mon: enable/disable mgr modules via 'ceph mgr module ...' commands

### PR DESCRIPTION
**Pull request by liewegas**
This doesn't include any magic for auto-enabling modules that are installed.  I'm
not sure that's such a good idea anyway.

### CODE REVIEW

**Comment by liewegas** [comment]
@tserong

**Comment by jcsp** [comment]
It's not essential, but I think this will be friendlier if we have the mgr daemons report their available plugins to the mons, so that we can warn (if not forbid) the user if they are enabling a nonexistent module -- typos etc.

**Comment by liewegas** [comment]
I don't really like the MgrMap structure of having the top level contain the active mgr fields and standbys map contain the others.  Maybe we can move to a more MDSMap-like structure where we have a map of all daemons, and a single top-level field indicating which one is the active one?

File: src/mgr/PyModules.cc [hunk 1]
```diff
@@ -664,3 +666,33 @@ PyObject *PyModules::get_context()
...
+	modules->insert(n);
+      }
+    }
+  }
+  closedir(dir);
+}
+
+void PyModules::list_modules(std::set<std::string> *modules)
+{
+  // FIXME search system paths too?
```
Comment (jcsp) [inline src/mgr/PyModules.cc]: Restricting to mgr_module_path is by design -- if we used system paths then our mgr module names would potentially collide with any ordinary python module names on the system.

File: src/mgr/PyModules.cc [hunk 2]
```diff
@@ -664,3 +666,33 @@ PyObject *PyModules::get_context()
...
+    return;
+  }
+  struct dirent *entry = NULL;
+  while ((entry = readdir(dir)) != NULL) {
+    string n(entry->d_name);
+    string fn = path + "/" + n;
+    struct stat st;
+    int r = ::stat(fn.c_str(), &st);
+    if (r == 0 && S_ISDIR(st.st_mode)) {
+      string initfn = fn + "/__init__.py";
```
Comment (jcsp) [inline src/mgr/PyModules.cc]: Should check for "module.py" -- we might have an __init__.py in a directory that was just e.g. shared code in a module.

The fixed module.py name is a bit hacky but probably actually helps in this case.

File: doc/mgr/administrator.rst [hunk 3]
```diff
@@ -76,12 +76,6 @@ OPTION(mgr_module_path, OPT_STR, CEPH_PKGLIBDIR "/mgr") // where to load python
 :Type: String
 :Default: ``"<library dir>/mgr"``
 
-``mgr modules``
-
-:Description: List of python modules to load
-:Type: String
-:Default: ``"rest"`` (Load the REST API module only)
```
Comment (tserong) [inline doc/mgr/administrator.rst]: Should be "restful"

File: src/mon/MgrMap.h [hunk 4]
```diff
@@ -36,17 +38,21 @@ class StandbyInfo
 
   void encode(bufferlist& bl) const
   {
-    ENCODE_START(1, 1, bl);
+    ENCODE_START(2, 1, bl);
```
Comment (tserong) [inline src/mon/MgrMap.h]: Is this inconsistent with the HEAD_VERSION in MMgrBeacon.h? Or do I just not understand this piece?

File: src/mon/MgrMap.h [hunk 5]
```diff
@@ -36,17 +38,21 @@ class StandbyInfo
...
     ENCODE_FINISH(bl);
   }
 
   void decode(bufferlist::iterator& p)
   {
-    DECODE_START(1, p);
+    DECODE_START(2, p);
     ::decode(gid, p);
     ::decode(name, p);
+    if (struct_v >= 2) {
```
Comment (tserong) [inline src/mon/MgrMap.h]: Same question as above

File: src/mon/MgrMap.h (hunk 4 above)
Comment (liewegas) [inline src/mon/MgrMap.h]: Unrelated... the 1 and 2 here are the version for this struct only (StandbyInfo)

File: doc/mgr/administrator.rst (hunk 3 above)
Comment (liewegas) [inline doc/mgr/administrator.rst]: Removed now anyway :)

**Comment by liewegas** [comment]
http://pulpito.ceph.com/sage-2017-06-30_05:44:03-rados-wip-sage-testing-distro-basic-smithi/

**Comment by liewegas** [comment]
@jcsp ok to merge do you think?

**Comment by jcsp** [comment]
:+1: no more fumbling around with mgr_modules!

**Pull request merged**